File: AtCoder/AGC027/B.cpp
#include <iostream>
#include <vector>

using namespace std;

int main(){
    long long N, X; cin >> N >> X;
    vector<long long> x(N);
    vector<long long> s(N+1, 0);
    for(auto& t : x) cin >> t;
    for(int i=0;i<N;i++) s[i+1] = s[i] + x[i];
    long long res = (1LL << 60);
    for(int i=N;i>=1;i--){
        long long cur = (N+i) * X;
        for(int j=0;j*i<N;j++){
            int src = N - j * i;
            int dst = max(0, src - i);
            cur += (s[src] - s[dst]) * max(5, 2*j+3);
            if(cur > res) break;
        }
        res = min(res, cur);
    }
    cout << res << endl;
}

File: AtCoder/AGC027/B.rs
use proconio::input;

fn main(){
    input! {
        n : usize,
        c : i128,
        x : [i128; n]
    };
    let mut s = vec![0; n+1];
    for i in 0..n { s[i+1] = s[i] + x[i]; }
    let mut res = 1 << 120;
    for i in 1..=n {
        let mut cur = (n+i) as i128 * c;
        for j in 0..=n {
            if i*j >= n { break }
            let src = n - j*i;
            let dst = if src >= i { src - i } else { 0 };
            cur += (s[src] - s[dst]) * std::cmp::max(5, 2*j+3) as i128;
        }
        res = std::cmp::min(res, cur);
    }
    println!("{}", res);
}